src/main/java/hudson/remoting/RemoteClassLoader.java

```diff
@@ -130,10 +135,21 @@ private RemoteClassLoader(@CheckForNull ClassLoader parent, @Nonnull IClassLoade
         final Channel channel = RemoteInvocationHandler.unwrap(proxy);
         this.channel = channel == null ? null : channel.ref();
         this.underlyingProxy = proxy;
+        final IClassLoader targetProxy;
         if (channel == null || !channel.remoteCapability.supportsPrefetch() || channel.getJarCache()==null) {
-            proxy = new DumbClassLoaderBridge(proxy);
+            targetProxy = new DumbClassLoaderBridge(proxy);
+        } else {
+            targetProxy = proxy;
+        }
+        if (MAX_RETRIES_ON_REMOTING_INTERRUPTION > 0) {
```

Comment:
There's a readability refactoring here that would be nice. I was confused on reading this and may not be the only one.

I propose an Extract Method refactoring that looks something like this:
```suggestion
        if (shouldRetryOnRemotingInterruption()) {
```

Reply:
505ad3b4 - I've extracted it as `static RemoteInterruptedExceptionCatcher.shouldBeEnabled()`. This way, all references to `MAX_RETRIES_ON_REMOTING_INTERRUPTION` are in `RemoteInterruptedExceptionCatcher`.

Reply:
I like it.

Reply:
Regarding this `RemoteClassLoader` constructor, we could refactor a little bit more and extract more static methods, something like that:

```java

    private static IClassLoader enhanceDumbClassLoader(@CheckForNull Channel channel, @Nonnull IClassLoader proxy) {
        if (channel == null || !channel.remoteCapability.supportsPrefetch() || channel.getJarCache() == null) {
            return new DumbClassLoaderBridge(proxy);
        } else {
            return proxy;
        }
    }

    private static IClassLoader addRemoteInterruptedExceptionCatcher(@Nonnull IClassLoader proxy) {
        if (RemoteInterruptedExceptionCatcher.shouldBeEnabled()) {
            // JENKINS-61103: wrap the target (IClassLoader) proxy in a dynamic proxy which will handle retries in case
            // of interruption exception from the remoting system during class loading/initialization.
            // It should take care of cases like JENKINS-36991, JENKINS-51854 and others.
            return (IClassLoader) Proxy.newProxyInstance(RemoteClassLoader.class.getClassLoader(),
                    new Class[] { IClassLoader.class }, new RemoteInterruptedExceptionCatcher(proxy));
        } else {
            return proxy;
        }
    }

    private RemoteClassLoader(@CheckForNull ClassLoader parent, @Nonnull IClassLoader proxy) {
        super(new URL[0],parent);
        final Channel channel = RemoteInvocationHandler.unwrap(proxy);
        this.channel = channel == null ? null : channel.ref();
        this.underlyingProxy = proxy;
        this.proxy = addRemoteInterruptedExceptionCatcher(enhanceDumbClassLoader(channel, proxy));
    }
```

Or maybe the same but keeping an intermediate `targetProxy` variable:

```java
        final IClassLoader targetProxy = enhanceDumbClassLoader(channel, proxy);
        this.proxy = addRemoteInterruptedExceptionCatcher(targetProxy);
```

I don't really have a preference, you let me know if you have one.

Reply:
I thought about some of those as I was reading through it and think it could help clarify what's going on. I don't know that it's necessary but it could be nice.